### PR TITLE
Add support for Testcontainers

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
When Testcontainers is chosen along with any of supported JDBC drivers it will add dependency to Testcontainers database module related to selected JDBC driver.

When Testcontainers is chosen along with R2DBC driver and Spring Data R2DBC it will add extra dependency to Testcontainers R2DBC module.

Relevant links to Testcontainers documentation are added to `HELP.md` file.

### CODE REVIEW

File: start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	private final Map<String, String> driverToTestContainersDependency = new HashMap<>();
+
+	private final InitializrMetadata metadata;
+
+	public TestcontainersBuildCustomizer(InitializrMetadata metadata) {
+		this.metadata = metadata;
+		this.driverToTestContainersDependency.put("postgresql", "postgresql");
+		this.driverToTestContainersDependency.put("mysql", "mysql");
+		this.driverToTestContainersDependency.put("sqlserver", "mssqlserver");
+		this.driverToTestContainersDependency.put("oracle", "oracle-xe");
```
Comment (michael-simons) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java:51]: There's also support for Neo4j Testcontainers… Just saying :)

Comment (simasch) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java:51]: And Cassandra, CockroachDB, Couchbase, Clickhouse, DB2, Dynalite, InfluxDB, MariaDB, MongoDB, MS SQL Server, OrientDB and Presto

Comment (maciejwalkowiak) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java:51]: Thanks for hints. Neo4j, Couchbase, Cassandra and MongoDB should be added here indeed. For the rest there is no option to choose them at `start.spring.io`.

**Comment by snicoll** [comment]
@maciejwalkowiak thanks for the PR. There are plans in Spring Boot to ease the integration with testcontainers which should provide dependency management for it and a number of utilities. That would be the logical step to add it here.

Flagging for team attention.

**Comment by wilkinsona** [comment]
I can't see any harm in adding it to start.spring.io now. We want to encourage the use of Testcontainers and this feels to me like a good step in that direction. I don't think it will inhibit any integration that we may add to Spring Boot in the future.

**Comment by snicoll** [comment]
Agreed, it won't. I was just thinking whether we could do something to go the extra mile (and the upcoming SB support could help with that). Offering an option and let the test fails is ok I guess but that's a bit unusual compared to what we've done in the past.

**Comment by maciejwalkowiak** [comment]
Support for Neo4j, Cassandra and Couchbase added as @michael-simons and @simasch suggested. Mongo has not been added as Testcontainers module for Mongo is in "experimental" state. Thanks for review @snicoll and @wilkinsona. I am happy to adjust it if you have any suggestions.

File: start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2012-2019 the original author or authors.
```
Comment (eddumelendez) [inline start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java:2]: 2020? the same for the other classes

**Comment by vasilievip** [comment]
@maciejwalkowiak I would like to add these modules as well, can I piggy back on your PR to do so, like add new commits?  
https://github.com/testcontainers/testcontainers-spring-boot

**Comment by bsideup** [comment]
@vasilievip let's have a separate discussion about it. With the recent changes in Spring Boot, the linked project's value is small but creates a bit of confusion for users who want to customize the containers and/or their lifecycle.

**Comment by vasilievip** [comment]
@bsideup yes, I agree that with latest changes in spring boot there is less code to write in order to get started, but you still need to write code (except jdbc compatible datastores which handled by testcontainers jdbc driver wrapper) while in linked project you can have everything up and running only by adding few deps w/o single line of code written. 
So, I would say spring boot support for testcontainers is not there yet. Lets talk in testcontainers slack.

**Comment by vasilievip** [comment]
Discussed with @bsideup to wait for spring boot improvements for testcontainers integration and hold with existing autoconfigurations

File: start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java (hunk 2 above)
Comment (maciejwalkowiak) [reply start-site/src/main/java/io/spring/start/site/extension/dependency/testcontainers/TestcontainersBuildCustomizer.java]: Thanks @eddumelendez good catch!

**Comment by snicoll** [comment]
Thanks @maciejwalkowiak, I've polished things up in https://github.com/spring-io/start.spring.io/commit/f55d58a51d8271006a642048a74edc45efa629d2

**Pull request closed without merging**